Added course metadata and enrollment summary tables.

dsjen opened the pull request:
I'll be adding acceptance tests next and this is hopefully in a good enough state for an early review at the very least.  Please feel free to review!

Changes include:
- In `load_internal_reporting_course_catalog`, wrapped Extract\* classes into `HivePartionTask` and created analogous `BareHiveTableTask`
- Added `course_enrollment_summary.py` that includes the job to create a course summary sql table with course metadata and recent enrollment data

@thallada @HassanJaveed84 @brianhw @mulby

brianhw commented on edx/analytics/tasks/course_enrollment_summary.py:
```diff
@@ -0,0 +1,203 @@
+                enrollment_end.count,
+                (enrollment_end.count - COALESCE(enrollment_start.count, 0)) as count_change_7_days
+                enrollment_end.cumulative_count,
+                enrollment_end.mode,
```
Hmm, now I'm confused.  Why does the order of these columns not match the order in the CourseEnrollmentSummaryRecord?  I would have thought that they needed to match.

dsjen replied:
I didn't realize!  I will make them match.

brianhw commented on edx/analytics/tasks/course_enrollment_summary.py:
```diff
@@ -0,0 +1,203 @@
+            LEFT JOIN {course_table} as course
+                ON course.course_id = ce_recent.course_id
+            WHERE ce_recent.date = '{end_date}'
+            AND ce_early.date = '{start_date}';
```
What is ce_recent and ce_early?

dsjen replied:
Left over from an earlier prototype.  ce_recent is enrollment_end and ce_early is enrollment_start.  Thanks for catching that!

brianhw commented on edx/analytics/tasks/course_enrollment_summary.py:
```diff
@@ -0,0 +1,203 @@
+
+    @property
+    def table(self):
+        return 'course_summary'
```
course_summary_enrollments?   Or course_enrollment_summary?  I'm confused by the sudden inversion in naming.

dsjen replied:
Ugh, I'm being sloppy on the naming.  I did noticed that course_enrollment_summary is used in [enrollments.py](https://github.com/edx/edx-analytics-pipeline/blob/457dde57faac923256752270ccca2e684c7a3eeb/edx/analytics/tasks/enrollments.py#L596), but inverting the summary and enrollments doesn't seem like a good solution.  What do you think if I called this `course_meta_summary_enrollment` instead?  Too weird?

dsjen commented on edx/analytics/tasks/course_enrollment_summary.py:
```diff
@@ -0,0 +1,203 @@
+
+from edx.analytics.tasks.util.hive import (
+    BareHiveTableTask,
+    hive_database_name,
```
I noticed it's used in [module_engagement](https://github.com/edx/edx-analytics-pipeline/blob/cc233ee70dfc40fd4f7d80d3825d2ccc21475acf/edx/analytics/tasks/module_engagement.py#L1169).  I could refactor the name to something more fitting--maybe too much for this PR though.

dsjen commented on edx/analytics/tasks/course_enrollment_summary.py:
```diff
@@ -0,0 +1,203 @@
+    """Recent enrollment summary and metadata for a course."""
+    course_id = StringField(length=255, nullable=False)
+    catalog_course_title = StringField(nullable=True, length=255,
+                                       normalize_whitespace=True)
```
Ok, I'll plan on adding descriptions to the fields and format as needed.

dsjen commented on edx/analytics/tasks/course_enrollment_summary.py:
```diff
@@ -0,0 +1,203 @@
+    end_time = DateTimeField(nullable=True)
+    count = IntegerField(nullable=True)
+    count_change_7_days = IntegerField(nullable=True)
+    cumulative_count = IntegerField(nullable=True)
```
Ah, thanks!  I'll reorder.

dsjen commented on edx/analytics/tasks/load_internal_reporting_course_catalog.py:
```diff
@@ -252,7 +293,48 @@ class CourseRecord(Record):
+        return CourseRecord.get_hive_schema()
+
+
+class ExtractCoursePartitionTask(LoadInternalReportingCourseCatalogMixin, HivePartitionTask):
```
I didn't see "extract" in the other parts of the codebase either and thought that it could be renamed--maybe just removing "extract" entirely as it's assumed.  Thoughts?

I'm sort of at a loss for the "course" naming conundrum.  What does it mean to be a program course vs. just a course?  One appears to have marketing information attached to it and the other program information.  Maybe:
* MarketingCourse
* ProgramCourse

Would it make sense to have an AbstractCourseRecord?

mulby replied:
I think just removing "Extract" is sufficient IMHO.

dsjen replied:
Ok, I'll plan on removing "Extract" unless there are additional thoughts on this!

mulby commented on edx/analytics/tasks/course_enrollment_summary.py:
```diff
@@ -0,0 +1,203 @@
+            database_name=hive_database_name(),
+            enrollment_table='course_enrollment_mode_daily',
+            program_table='program_course',
+            course_table='course_catalog',
```
I don't see a huge amount of value in pulling these table names out into the format string unless you were to actually dynamically fetch the table names from the requirements. We usually just hardcode table names in queries like this one (for better or worse)!

dsjen replied:
Gotcha!

dsjen commented on edx/analytics/tasks/course_enrollment_summary.py:
```diff
@@ -0,0 +1,203 @@
+
+
+class CourseEnrollmentSummaryWrapperTask(CourseEnrollmentSummaryDownstreamMixin,
+                                         luigi.WrapperTask):
```
I was feeling a little iffy on this class too as it's not needed.  If I remember correctly, @brianhw had suggested to me that having a wrapper class as the entry point would be helpful in case other workflow were to be added.  Is there a pattern that's typically done?  Is this preemptive optimization?

brianhw commented on edx/analytics/tasks/course_summary_enrollment.py:
```diff
@@ -0,0 +1,232 @@
+    pass
+
+
+class CourseSummaryEnrollmentWrapperTask(CourseSummaryEnrollmentDownstreamMixin,
```
We also mark entry points with a decorator:  @workflow_entry_point

brianhw replied:
Also not sure about how important, but I think we generally define the classes in the order they are used.  So since this uses ImportCourseSummaryEnrollmentsIntoMysql, it would normally appear after it.  (I think it's required if it inherits, but less so if the dependency is in a function.)

brianhw commented on edx/analytics/tasks/course_summary_enrollment.py:
```diff
@@ -0,0 +1,232 @@
+
+class CourseSummaryEnrollmentWrapperTask(CourseSummaryEnrollmentDownstreamMixin,
+                                         luigi.WrapperTask):
+    """Entrypoint for course summary enrollment task."""
```
nit: I would suggest making this two words:  "Entry point"....

brianhw commented on edx/analytics/tasks/course_summary_enrollment.py:
```diff
@@ -0,0 +1,232 @@
+                enrollment_mode,
+                count,
+                count_change_7_days,
+                cumulative_count
```
Hmm.  This is all the fields in the CourseSummaryEnrollmentRecord, right?  I wonder if there's an easy way to extract the list from there.  Something like ','.join([field.name for field in CourseSummaryEnrollmentRecord.fields()]).  Just to make it DRYer.

brianhw replied:
Or one could just get it from the self.columns, which already gets it from the record's get_sql_schema.  Just extract the names.

dsjen replied:
Oh, I like that!  Will do.

brianhw commented on edx/analytics/tasks/course_summary_enrollment.py:
```diff
@@ -0,0 +1,232 @@
+            date=self.date,
+            api_root_url=self.api_root_url,
+            api_page_size=self.api_page_size,
+            overwrite=self.overwrite,
```
Do you have a sense of the scope of the overwrite as opposed to the overwrite_n_days?  I'm assuming the former is for catalog, and the latter for enrollment?  (Possibly worth a comment somewhere in that class?  Or where?)

dsjen replied:
Right, `overwrite_n_days` is for enrollment and `overwrite` is for the catalog.  I'll add relevant comments.  Thanks!

dsjen commented:
Hooray!  Acceptance tests for test_course_summary_enrollment and test_enrollments (because I changed some of the enrollment event fixtures) pass and `make coverage` passes.  I'm running the entire acceptance test suite currently.

To recap, this PR includes:
- In `load_internal_reporting_course_catalog`, wrapped Extract\* classes into `HivePartionTask` and created analogous `BareHiveTableTask`
- Added `course_enrollment_summary.py` that includes the job to create a course summary sql table with course metadata and recent enrollment data

The job can be run with a parameter to turn off population of the course catalog tables for the case that the catalog service isn't available.

Please feel free to review!  I'll be sure to squash commits once approved.

@thallada @HassanJaveed84 @brianhw @mulby

brianhw commented on edx/analytics/tasks/course_summary_enrollment.py:
```diff
@@ -0,0 +1,171 @@
+            ),
+        ]
+
+        if self.disable_course_catalog:
```
This is a nice clean solution.  It might warrant a brief comment.

dsjen replied:
Can do!

brianhw commented on edx/analytics/tasks/tests/acceptance/test_course_summary_enrollment.py:
```diff
@@ -0,0 +1,95 @@
+    CATALOG_DATE = '2016-09-08'
+
+    def setUp(self):
+        ''' Loads enrollment and course catalog fixtures. '''
```
nit:  Using single quotes followed by a space looks cool, but it's different than our standard.

dsjen commented:
I believe all acceptance tests should be fixed now.  I'm running the entire suite now.  Please review when you have a moment.

dsjen commented:
FYI, acceptance tests have all passed!

mulby commented on edx/analytics/tasks/course_summary_enrollment.py:
```diff
@@ -0,0 +1,173 @@
+    @property
+    def query(self):
+        end_date = self.interval.date_b - datetime.timedelta(days=1)
+        start_date = self.interval.date_b - datetime.timedelta(days=7)
```
don't we want this to be `end_date - datetime.timedelta(days=7)`?

dsjen replied:
Good catch!  I think I was confused.  I was thinking that lets says that `date_b` is 2016-11-17.  `end_date` would then be 2016-11-16 and the `start_date` would be 2016-11-10 for a total of 7 complete days.

However, the ask was to be able to essentially compare yesterday's enrollment counts to the same day, e.g. compare Monday and the previous Monday.  I'll update this!

mulby commented on edx/analytics/tasks/course_summary_enrollment.py:
```diff
@@ -0,0 +1,173 @@
+                AND enrollment_start.mode = enrollment_end.mode
+                AND enrollment_start.date = '{start_date}'
+            LEFT OUTER JOIN program_course program
+                ON program.course_id = enrollment_end.course_id
```
Note that courses can belong to multiple programs, so this could result in multiple rows in the output for each course - is that the desired behavior? I suppose we could handle the display logic downstream if we wanted to consolidate in the API or UI tiers...

mulby replied:
It's actually a kind of hard problem, you generally need to figure out what to do when a course belongs to multiple programs - should we just use the first one alphabetically? Should the programs be presented as a comma separated list?

dsjen replied:
Oh, that's very interesting.  I guess it might be confusing to the user to see just the first program, and your suggestion of turning them into a comma separated list would help solve that.  I can change the query so that both `catalog_course_title` and `catalog_course` from from the course_catalog table instead, which then leaves a group by (right?) for the `program_id` and `program_title`.

The program data isn't needed for the MVP course listing table in Insights, so maybe we can punt on including the program data for now.  Opinions?

@katymyw - is it important that we include `program_id` and `program_title` in the course listing table in the resultstore?  It's not in the wireframes, but perhaps you had some forward thinking reason for wanting it available.

katymyw replied:
@dsjen -- this was a little more forward-thinking (e.g., lightweight program dashboard could be just filtering to the courses in a program). In which case we would want to know all programs a course is in. But I'd say scrap it since it's not immediately needed and adds complexity here.

mulby commented on edx/analytics/tasks/course_summary_enrollment.py:
```diff
@@ -0,0 +1,173 @@
+
+        # if disable_course_catalog is set, create the hive tables w/o loading data
+        if self.disable_course_catalog:
+            catalog_tasks = [task.hive_table_task for task in catalog_tasks]
```
nit: reassignment of the `catalog_tasks` variable here was a little confusing to me... I think something like this is more clear:

```python
if self.disable_course_catalog:
    # this Open edX installation isn't running a catalog service, so just create empty hive tables without loading any data into them
    yield [task.hive_table_task for task in catalog_tasks]
else:
    # create the hive tables and populate them with data
    yield catalog_tasks
```

dsjen replied:
Makes sense!

mulby commented on edx/analytics/tasks/course_summary_enrollment.py:
```diff
@@ -0,0 +1,173 @@
+
+    @property
+    def required_table_tasks(self):
+        yield EnrollmentByModeTask(
```
Historically we have referenced other tables like this using external URLs since this could cause some wacky race conditions if this workflow's schedule overlapped with the enrollment workflow.

We may want to carefully consider how these three workflows are related and when each one needs to run relative to one another:

1. The workflow that pulls the catalog API data into the warehouse.
2. The enrollment workflow
3. This new workflow

I see potential here for them to stomp on one another depending on how they are scheduled.

dsjen commented:
Unit tests, linter, and acceptance tests on enrollment have passed successfully.  I'm running the full acceptance test suite now.  Please leave any additional feedback ASAP!

brianhw commented on edx/analytics/tasks/enrollments.py:
```diff
@@ -859,24 +866,145 @@ def columns(self):
+
+class CourseSummaryEnrollmentRecord(Record):
+    """Recent enrollment summary and metadata for a course."""
+    course_id = StringField(length=255, nullable=False, description='A unique identifier of the course')
```
tiny-nit:  this is the only one with length before nullable.

dsjen replied:
Done!

brianhw commented on edx/analytics/tasks/enrollments.py:
```diff
@@ -859,24 +866,145 @@ def columns(self):
+
+    disable_course_catalog = luigi.BooleanParameter(
+        config_path={'section': 'course-summary-enrollment', 'name': 'disable_course_catalog'},
+        default=False,
```
I think this should be true by default.  Or we change the sense of it, and call it "enable_course_catalog" with a default of False.  I think most people outside of edX who will be running enrollment will not have a course catalog set up.   Once we have something better documented for helping them do the setup or set the flag, then we could change the default in the code if we really wanted to.  But I think it would be pretty disruptive to have people run enrollment and have to find out about setting this flag.

dsjen replied:
Ok, I'll update!

dsjen commented:
@brianhw -- reran the enrollment acceptance tests with your requested changes and they've passed.  I'm rerunning the entire acceptance suite with the change now.

dsjen commented:
The test_enrollment_validation acceptance test passed.  I would be shocked if the run of the entire acceptance suite doesn't pass.  Please chime in if there's anything else!

dsjen commented:
Acceptance tests have passed!  Merging soon.
